src/plugins/intel_gpu/include/intel_gpu/primitives/data.hpp

```diff
@@ -74,12 +74,47 @@ struct data : public primitive_base<data> {
         if (_allocation_type == allocation_type::usm_host || _allocation_type == allocation_type::usm_shared) {
             ib >> make_data(mem->buffer_ptr(), data_size);
         } else {
-            std::vector<uint8_t> _buf;
-            _buf.resize(data_size);
-            ib >> make_data(_buf.data(), data_size);
-            // stream* strm = reinterpret_cast<stream*>(ib.get_stream());
+            const size_t DATA_BLOCK_SIZE = 2 * 1024 * 1024;
             auto& strm = ib.get_engine().get_service_stream();
-            mem->copy_from(strm, _buf.data());
+            if (data_size < DATA_BLOCK_SIZE || output_layout.format.is_image_2d()) {
+                std::vector<uint8_t> _buf(data_size);
+                ib >> make_data(_buf.data(), data_size);
+                mem->copy_from(strm, _buf.data());
+            } else {
+                std::vector<uint8_t> _buf1(DATA_BLOCK_SIZE);
+                std::vector<uint8_t> _buf2(DATA_BLOCK_SIZE);
+                bool buf_flag = true;
+                event::ptr ev1, ev2;
+                ev1 = ev2 = nullptr;
+
+                size_t dst_offset = 0;
+                while (dst_offset < data_size) {
+                    size_t copy_size = (data_size > (dst_offset + DATA_BLOCK_SIZE)) ? DATA_BLOCK_SIZE : (data_size - dst_offset);
+                    if (buf_flag) {
+                        ib >> make_data(_buf1.data(), copy_size);
+                        if (ev2 != nullptr) {
+                            ev2->wait();
+                            ev2 = nullptr;
+                        }
+                        ev1 = mem->copy_from(strm, _buf1.data(), false, dst_offset, copy_size);
+                    } else {
+                        ib >> make_data(_buf2.data(), copy_size);
+                        if (ev1 != nullptr) {
+                            ev1->wait();
+                            ev1 = nullptr;
+                        }
+                        ev2 = mem->copy_from(strm, _buf2.data(), false, dst_offset, copy_size);
+                    }
+                    dst_offset += DATA_BLOCK_SIZE;
+                    buf_flag = !buf_flag;
+                }
+                if (ev2 != nullptr) {
+                    ev2->wait();
+                }
+                if (ev1 != nullptr) {
+                    ev1->wait();
```

Comment:
Will it be impact on performance, if we collect all event in this loop and then wait for them out of the loop?

Reply:
The host buffer memory buffer should also be maintained until the related event is completed, which is likely to increase memory usage.